// angular-auth-oidc-client/esm2022/lib/public-events/event-types.mjs
// eslint-disable-next-line no-shadow
export var EventTypes;
(function (EventTypes) {
    /**
     *  This only works in the AppModule Constructor
     */
    EventTypes[EventTypes["ConfigLoaded"] = 0] = "ConfigLoaded";
    EventTypes[EventTypes["CheckingAuth"] = 1] = "CheckingAuth";
    EventTypes[EventTypes["CheckingAuthFinished"] = 2] = "CheckingAuthFinished";
    EventTypes[EventTypes["CheckingAuthFinishedWithError"] = 3] = "CheckingAuthFinishedWithError";
    EventTypes[EventTypes["ConfigLoadingFailed"] = 4] = "ConfigLoadingFailed";
    EventTypes[EventTypes["CheckSessionReceived"] = 5] = "CheckSessionReceived";
    EventTypes[EventTypes["UserDataChanged"] = 6] = "UserDataChanged";
    EventTypes[EventTypes["NewAuthenticationResult"] = 7] = "NewAuthenticationResult";
    EventTypes[EventTypes["TokenExpired"] = 8] = "TokenExpired";
    EventTypes[EventTypes["IdTokenExpired"] = 9] = "IdTokenExpired";
    EventTypes[EventTypes["SilentRenewStarted"] = 10] = "SilentRenewStarted";
    EventTypes[EventTypes["SilentRenewFailed"] = 11] = "SilentRenewFailed";
})(EventTypes || (EventTypes = {}));
